s64da_benchmark_toolkit/reporting.py

```diff
@@ -0,0 +1,172 @@
+
+        while not reporting_queue.empty():
+            query_metric = reporting_queue.get()
+
+            if self.explain_analyze:
+                self._save_explain_plan(query_metric)
+
+            if self.scale_factor:
+                self._save_query_output(query_metric)
+
+            df = df.append(query_metric.dataframe)
+
+        df = df.reset_index(drop=True)
+
+        netdata_config = self.config.get('netdata')
+        if netdata_config:
+            netdata = Netdata(netdata_config)
+            if self.netdata_output_file:
+                netdata.write_stats(df, self.netdata_output_file)
+
+        if self.scale_factor:
+            df = self._check_correctness(df)
+
+        total_runtime = df['timestamp_stop'].max() - df['timestamp_start'].min()
+        total_runtime_seconds = total_runtime.total_seconds()
+        self._print_results(df, total_runtime_seconds)
+
+        print(f'\nTotal runtime: {total_runtime} ({total_runtime_seconds:.2f}s)')
+
+    def _save_explain_plan(self, query_metric):
+        plan_file_name = query_metric.make_file_name('txt')
+        plan_file_path = os.path.join(self.explain_analyze_dir, plan_file_name)
+        os.makedirs(os.path.dirname(plan_file_path), exist_ok=True)
+
+        with open(plan_file_path, 'w') as plan_file:
+            plan_file.write(query_metric.plan)
+
+    def _save_query_output(self, query_metric):
+        query_result = query_metric.result
+        if query_result is not None and query_result[1]:
+            query_result_header = query_result[0]
+            query_result_data = query_result[1]
+
+        else:
+            query_result_header = []
+            query_result_data = []
+
+        csv_file_name = query_metric.make_file_name('csv')
+        csv_file_path = os.path.join(self.query_results, csv_file_name)
+        os.makedirs(os.path.dirname(csv_file_path), exist_ok=True)
+        with open(csv_file_path, 'w') as csv_file:
+            csvfile = csv_writer(csv_file)
+            csvfile.writerow(query_result_header)
+            csvfile.writerows(query_result_data)
+
+    @classmethod
+    def _sort_df(cls, df):
+        index_sort = index_natsorted(zip(df['stream_id'], df['query_id']))
+        df = df.reindex(index=order_by_index(df.index, index_sort))
+        df = df.reset_index(drop=True)
+        return df
+
+    def _print_results(self, df, total_runtime_seconds):
+        df = Reporting._sort_df(df)
+
+        if 'print' in self.output:
+            print('\n')
+            print(tabulate(df, headers='keys', tablefmt='github', floatfmt='.2f'))
+
+        if 'csv' in self.output:
+            if self.csv_file:
+                df.to_csv(self.csv_file, sep=';')
+
+        if 'correctness_check' in df:
+            report_datetime = datetime.now().strftime('%d-%m-%Y %H:%M:%S')
+            with open(self.html_output, 'w') as html_output_file:
+                html_output_file.write(f'''
+<h1>Swarm64 Benchmark Results Report</h1>
```

Comment:
Here's another place where HTML representation is mixed with the code. IMHO, it should be extracted.

Reply:
Good point, but this is for compatibility with the current implementation to not break si-tpc-toolkit. What would be your recommendation?

Reply:
I would put the HTML code in a jinja template and render it here.

Reply:
Since this is only this tiny bit of template, I thought it would be acceptable to not have that overhead.

Reply:
I'll approve the PR also if this stays, but I would rather see it extracted :)

Reply:
This results in a mix of pandas rendering + Jinja, not worth the extra overhead IMO.

Reply:
ok, for now. Should probably be revisited in the future though.